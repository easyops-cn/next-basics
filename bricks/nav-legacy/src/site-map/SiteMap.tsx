import React, { useState } from "react";
import { AutoComplete, Input, Popover, Tooltip } from "antd";
import styles from "./SiteMap.module.css";
import { IHeaderNav } from "./interface";
import { getNavModuleList } from "./nav";
import {
  getFilterNavList,
  setFlags,
  COLUMNS,
  getGroupsList,
} from "./nav/utils";
import { orderBy, keyBy } from "lodash";
import { GeneralIcon } from "@next-libs/basic-components";
import { getHistory, getRuntime } from "@next-core/brick-kit";
import { isIframe } from "./nav/utils";
import { K, NS_NAV_LEGACY } from "../i18n/constants";
import { useTranslation } from "react-i18next";

interface SiteMapProps {
  modelMap: Record<string, any>[];
}

export function SiteMap(props: SiteMapProps): React.ReactElement {
  const flags = getRuntime().getFeatureFlags();
  setFlags(flags);

  const { modelMap } = props;
  const [visible, setVisible] = useState(true);
  const { t } = useTranslation(NS_NAV_LEGACY);

  const showModal = () => {
    setVisible(true);
  };

  const headerNavs: IHeaderNav[] = getNavModuleList(modelMap);
  const autoDataSource: { value: string; text: string }[] =
    getFilterNavList(headerNavs);
  const [dataSource, setDataSource] = useState(autoDataSource);

  const onSearch = (value) => {
    const q = !value ? "" : value.trim().toLowerCase();
    const filterData: { value: string; text: string }[] = !q
      ? [...autoDataSource]
      : autoDataSource.filter(
          (item) =>
            item.text.toLowerCase().includes(q) ||
            item.objectId.toLowerCase().includes(q)
        );

    setDataSource(filterData);
  };

  const map = keyBy(dataSource, "text");
  const onSelect = (text: any) => {
    if (isIframe) {
      window.open(map[text].value, "_parent");
      return;
    }
    window.open(map[text].value, "_self");
  };

  const groupsList = getGroupsList(
    COLUMNS,
    orderBy(headerNavs[0].groups, ["order"], ["asc"])
  );
  const child = dataSource.map((item) => (
    <AutoComplete.Option key={item.value} value={item.text}>
      {item.text}
    </AutoComplete.Option>
  ));
  const content = (
    <div className={styles.mainContainer}>
      <div className={styles.searchContainer}>
        <div className={styles.searchWrapper}>
          <AutoComplete
            style={{ width: 200 }}
            dropdownStyle={{ width: 300 }}
            onSelect={onSelect}
            onSearch={onSearch}
            placeholder={t(K.SITE_MAP_SEARCH_PLACEHOLDER)}
            optionLabelProp="value"
          >
            {child}
          </AutoComplete>
        </div>
      </div>
      <div className={styles.masonry}>
        {groupsList
          .filter((item) => item.length)
          .map((groups, i) => (
            <div style={{ paddingLeft: 5 }} className={styles.column} key={i}>
              {groups.map((subCategory, index: number) => (
                <div key={index} className={styles.item}>
                  <div className={styles.subCategoryContainer}>
                    <span className={styles.diamondWrapper}></span>
                    <span>{subCategory.category}</span>
                  </div>
                  {subCategory.states.map((item, iIndex: number) => (
                    <div className={styles.itemContainer} key={iIndex}>
                      {
                        <a
                          target={isIframe ? "_parent" : "_self"}
                          href={item.stateName}
                        >
                          {item.text}
                          {item.id === "brickGroup" && (
                            <span className={styles.brickGroupContainer}>
                              <GeneralIcon
                                icon={{
                                  lib: "easyops",
                                  category: "app",
                                  icon: "brick-group",
                                }}
                              />
                            </span>
                          )}
                        </a>
                      }
                    </div>
                  ))}
                </div>
              ))}
            </div>
          ))}
      </div>
    </div>
  );

  return (
    <div>
      <div className={styles.siteMap}>
        <div className={styles.siteMapTitle}>资源</div>
        <Popover
          placement="bottomRight"
          title={null}
          content={content}
          trigger="click"
        >
          <Tooltip placement="top" title="查看所有">
            <a className={styles.showIcon} onClick={showModal}>
              <GeneralIcon
                icon={{
                  lib: "easyops",
                  category: "app",
                  icon: "all-cmdb",
                }}
              />
            </a>
          </Tooltip>
        </Popover>
      </div>
    </div>
  );
}
