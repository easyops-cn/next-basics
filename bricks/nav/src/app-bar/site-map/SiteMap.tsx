import React, { forwardRef, useEffect, useMemo } from "react";
import styles from "./SiteMap.module.css";
import { Link } from "@next-libs/basic-components";
import { MicroApp } from "@next-core/brick-types";
import { launchpadService } from "../LaunchpadService";

interface SiteCategory {
  name: string;
  id: string;
  order: string;
  apps: {
    id?: string;
    sort?: string;
    name?: string;
    homepage?: string;
  }[];
}
export interface SiteMapProps {
  categoryList: SiteCategory[];
  containerStyle?: React.CSSProperties;
  onLoad?: () => void;
}

export function LeacySiteMap(
  props: SiteMapProps,
  ref: any
): React.ReactElement {
  const handleWheel = (e: React.WheelEvent) => {
    e.stopPropagation();
  };

  useEffect(() => {
    props.onLoad?.();
  }, []);

  const handleClick = (app: MicroApp): void => {
    launchpadService.pushVisitor("app", app);
  };

  return (
    <div
      className={styles.scrollContainer}
      style={props.containerStyle}
      ref={ref}
      onWheel={handleWheel}
    >
      <div className={styles.siteMapContainer}>
        {props.categoryList?.map((item) => (
          <div className={styles.groupWrapper} key={item.name}>
            <div>{item.name}</div>
            <ul className={styles.group}>
              {item.apps?.map((row, index) => (
                <li className={styles.item} key={index}>
                  <Link
                    to={row.homepage}
                    onClick={() => handleClick(row as MicroApp)}
                  >
                    {row.name}
                  </Link>
                </li>
              ))}
            </ul>
          </div>
        ))}
      </div>
    </div>
  );
}

export const SiteMap = forwardRef(LeacySiteMap);
