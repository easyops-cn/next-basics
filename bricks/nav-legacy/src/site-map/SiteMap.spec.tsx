import React from "react";
import { shallow } from "enzyme";
import { SiteMap } from "./SiteMap";
import { getRuntime } from "@next-core/brick-kit";

jest.mock("@next-core/brick-kit");

const mockGetRuntime = getRuntime as jest.Mock;
mockGetRuntime.mockReturnValue({
  getFeatureFlags: () => {
    return {
      "dongyin-flow": true,
    };
  },
});

const props = {
  modelMap: [
    {
      objectId: "APP",
      name: "app",
      category: "A",
    },
    {
      objectId: "HOST",
      name: "host",
      category: "B",
    },
  ],
};
describe("SiteMap", () => {
  it("should work", () => {
    const wrapper = shallow(<SiteMap {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
