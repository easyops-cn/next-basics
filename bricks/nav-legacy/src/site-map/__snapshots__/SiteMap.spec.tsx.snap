// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiteMap should work 1`] = `
<div>
  <div
    className="siteMap"
  >
    <div
      className="siteMapTitle"
    >
      资源
    </div>
    <Popover
      content={
        <div
          className="mainContainer"
        >
          <div
            className="searchContainer"
          >
            <div
              className="searchWrapper"
            >
              <ForwardRef(AutoComplete)
                dropdownStyle={
                  Object {
                    "width": 300,
                  }
                }
                onSearch={[Function]}
                onSelect={[Function]}
                optionLabelProp="value"
                placeholder="SITE_MAP_SEARCH_PLACEHOLDER"
                style={
                  Object {
                    "width": 200,
                  }
                }
              >
                <Option
                  value="A / app"
                >
                  A / app
                </Option>
                <Option
                  value="B / host"
                >
                  B / host
                </Option>
              </ForwardRef(AutoComplete)>
            </div>
          </div>
          <div
            className="masonry"
          >
            <div
              className="column"
              style={
                Object {
                  "paddingLeft": 5,
                }
              }
            >
              <div
                className="item"
              >
                <div
                  className="subCategoryContainer"
                >
                  <span
                    className="diamondWrapper"
                  />
                  <span>
                    A
                  </span>
                </div>
                <div
                  className="itemContainer"
                >
                  <a
                    href="/next/legacy/cmdb-instance-management/APP/list"
                    target="_self"
                  >
                    app
                  </a>
                </div>
              </div>
            </div>
            <div
              className="column"
              style={
                Object {
                  "paddingLeft": 5,
                }
              }
            >
              <div
                className="item"
              >
                <div
                  className="subCategoryContainer"
                >
                  <span
                    className="diamondWrapper"
                  />
                  <span>
                    B
                  </span>
                </div>
                <div
                  className="itemContainer"
                >
                  <a
                    href="/next/legacy/cmdb-instance-management/HOST/list"
                    target="_self"
                  >
                    host
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      }
      mouseEnterDelay={0.1}
      mouseLeaveDelay={0.1}
      overlayStyle={Object {}}
      placement="bottomRight"
      title={null}
      transitionName="zoom-big"
      trigger="click"
    >
      <Tooltip
        arrowPointAtCenter={false}
        autoAdjustOverflow={true}
        mouseEnterDelay={0.1}
        mouseLeaveDelay={0.1}
        placement="top"
        title="查看所有"
        transitionName="zoom-big-fast"
      >
        <a
          className="showIcon"
          onClick={[Function]}
        >
          <GeneralIcon
            icon={
              Object {
                "category": "app",
                "icon": "all-cmdb",
                "lib": "easyops",
              }
            }
          />
        </a>
      </Tooltip>
    </Popover>
  </div>
</div>
`;
